tests/func/test_to_database.py

```diff
@@ -0,0 +1,778 @@
+    }
+
+    chain.to_database(
+        "dots_format_table", connection, column_mapping=column_mapping_dots
+    )
+
+    # Test database format (double underscores) - this should also work
+    column_mapping_underscores = {
+        "id": "record_id",
+        "simple_field": "simple",
+        "nested_data__value": "nested_val",
+        "nested_data__config": "nested_cfg",
+    }
+
+    chain.to_database(
+        "underscores_format_table",
+        connection,
+        column_mapping=column_mapping_underscores,
+    )
+
+    engine = _get_engine_from_connection(connection)
+    with engine.connect() as conn:
+        result_dots = conn.execute(
+            sqlalchemy.text(
+                "SELECT record_id, simple, nested_val FROM dots_format_table "
+                "ORDER BY record_id"
+            )
+        )
+        rows_dots = result_dots.fetchall()
+
+        result_underscores = conn.execute(
+            sqlalchemy.text(
+                "SELECT record_id, simple, nested_val FROM underscores_format_table "
+                "ORDER BY record_id"
+            )
+        )
+        rows_underscores = result_underscores.fetchall()
+
+        # Both should have identical data
+        assert len(rows_dots) == 2
+        assert len(rows_underscores) == 2
+        assert rows_dots == rows_underscores
+        assert rows_dots[0][:3] == (1, "test1", "val1")
+        assert rows_dots[1][:3] == (2, "test2", "val2")
+
+
+def test_to_database_invalid_on_conflict_value(connection, test_session):
+    """Test that invalid on_conflict values raise ValueError."""
+    chain = dc.read_values(
+        id=[1, 2, 3],
+        name=["Alice", "Bob", "Charlie"],
+        session=test_session,
+    )
+
+    with pytest.raises(ValueError, match="on_conflict must be 'ignore' or 'update'"):
+        chain.to_database("test_table", connection, on_conflict="invalid")
+
+
+def test_to_database_with_null_values(connection, test_session, warehouse):
+    """Test to_database handles NULL values correctly."""
+    from datachain.data_storage.sqlite import SQLiteWarehouse
+
+    chain = dc.read_values(
+        id=[1, 2, 3, 4],
+        name=["Alice", None, "Charlie", "Diana"],
+        age=[25, 30, None, 28],
+        session=test_session,
+    )
+
+    chain.to_database("null_table", connection)
+
+    engine = _get_engine_from_connection(connection)
+    with engine.connect() as conn:
+        result = conn.execute(sqlalchemy.text("SELECT * FROM null_table ORDER BY id"))
+        rows = result.fetchall()
+
+    # Different warehouse implementations handle NULL values differently
+    default_str_value = None if isinstance(warehouse, SQLiteWarehouse) else ""
+    default_int_value = None if isinstance(warehouse, SQLiteWarehouse) else 0
+
+    assert len(rows) == 4
+    assert rows[0] == (1, "Alice", 25)
+    assert rows[1] == (2, default_str_value, 30)
+    assert rows[2] == (3, "Charlie", default_int_value)
+    assert rows[3] == (4, "Diana", 28)
```

Comment:
```suggestion
    assert list(rows) == [(1, "Alice", 25),
   (2, default_str_value, 30),
   (3, "Charlie", default_int_value),
(4, "Diana", 28)
]
```

Something like this. I am typing from mobile so formatting may be off